### PR TITLE
types: Add Expression Typings to Pipeline Stages

### PR DESCRIPTION
**Pull request by Uzlopak**
To be implemented:
- [x] Arithmetic Expression Operators
- [x] Array Expression Operators
- [x] Boolean Expression Operators
- [x] Comparison Expression Operators
- [x] Conditional Expression Operators
- [x] Custom Aggregation Expression Operators
- [x] Data Size Operators
- [x] Date Expression Operators
- [x] Literal Expression Operator
- [x] Miscellaneous Operators
- [x] Object Expression Operators
- [x] Set Expression Operators
- [x] String Expression Operators
- [x] Text Expression Operator
- [x] Trigonometry Expression Operators
- [x] Type Expression Operators
- [x] Accumulators
- [x] Variable Expression Operators
- [x] Window Operators

### CODE REVIEW

**Comment by Uzlopak** [comment]
I guess if I do every day about 20-30 Expressions than in 4-5 days this big chunk of code is done. 

Then we can use them for the PipelineStages and Queries. Should be better than the any typings we have currently.

Also there is I think a typescript feature were you can build attribute names based on strings to do stuff like key.$ where key is a key in an interface. But I think this could result in annoying issues like somebody using older typescript could not use the typings or you have keys after a projection stage were you don't have the typings of. 
So probably better to just use string as type and don't make it too intelligent.

Anyway, I guess, the best benefit will be when the pipeline stages have better expression support, were you know which fields are existing in an expression.

So this will be the goal of this PR.

**Comment by Uzlopak** [comment]
I should Copy more Of the documentation into the jsdoc.

**Review by vkarpov15 — COMMENTED** [review]
Couple of minor comments, but LGTM in general. Worst comes to worst re: casting, we just add some `any`, but I think this change is great for autocomplete

File: types/OlsonTimezoneCode.d.ts [hunk 1]
```diff
@@ -0,0 +1,599 @@
+/**
+ * tzDatabase 2021e
+ */
+export type OlsonTimezoneCode =
```
[the remark this answers is not in the document]
Comment (Uzlopak) [reply types/OlsonTimezoneCode.d.ts:4]: Tbh I also thought about implications and consider them neglectable.

I would like to include it. I didnt find any useful timezone definition npm package. 

To be honest, the timezone database is rarely changed. And the most used timezones are anyway basically never changed. Like Europe/Amsterdam or Europe/Berlin. It is a quite little typing feature. Also I think that the most changes to the database are for our use case more cosmetical nature, like changing the underlying timezone (GMT+2 to GMT+3), changing the timezones DST dates, changing linking and aliases for timezones etc.. 

2021e
https://en.wikipedia.org/w/index.php?title=List_of_tz_database_time_zones&diff=1052212854&oldid=1051985881
replaced Ukraine/Kyiv with Ukraine/Kiew but I think this was a political change by a ukrainian-wikipedian to use ukrainian version of Kiew for local patriotism. 
2021d
https://en.wikipedia.org/w/index.php?title=List_of_tz_database_time_zones&diff=1050116833&oldid=1050099921
irrelevant changes
2021c
https://en.wikipedia.org/w/index.php?title=List_of_tz_database_time_zones&diff=1047681771&oldid=1046406500
irrelevant changes
2021b
https://en.wikipedia.org/w/index.php?title=List_of_tz_database_time_zones&diff=1046346322&oldid=1045259187
irrelevant changes
2021a
https://en.wikipedia.org/w/index.php?title=List_of_tz_database_time_zones&diff=1004552847&oldid=998344043
irrelevant changes

and so on.

So basically the timezones didnt change for atleast a year.

Or else I have  type it as basic string, resulting that you can write anything you want into the timezone field. And thus making it quite fancyless. But if somebody really comes with the report that we have a missing timezone, we can always update the whole file based on the Database and should not accept partial updates like "Hey you are missing one timezone", but "hey here is a complete update of the timezone list"

File: types/Expression.d.ts [hunk 2]
```diff
@@ -0,0 +1,1959 @@
...
+
+    export type BooleanExpression =
+        boolean |
+        Path |
+        BooleanExpressionOperator |
+        ArrayExpressionOperatorReturningAny |
+        ComparisonExpressionOperatorReturningBoolean |
+        StringExpressionOperatorReturningBoolean;
+
+    export type NumberExpression =
```
[the remark this answers is not in the document]
Comment (Uzlopak) [reply types/Expression.d.ts:1730]: I think that we will use this for now primarely for aggregations and not for queries. But I can modify them to have a "strict" flag as generic type, which could make it less strict, if set to false.

[the remark this answers is not in the document]
Comment (Uzlopak) [reply types/Expression.d.ts]: We only use them for aggregation pipelines, which do not cast

**Comment by Uzlopak** [comment]
@vkarpov15 
Can this be part of 6.4?

File: .eslintrc.json [hunk 3]
```diff
@@ -4,43 +4,70 @@
   ],
   "ignorePatterns": [
     "docs",
+    "tools",
     "dist",
-    "test/files/*"
+    "website.js",
+    "test/files/*",
+    "benchmarks"
```
Comment (vkarpov15) [inline .eslintrc.json:11]: Do we really want to avoid linting benchmarks and `website.js`? Doesn't seem like the right thing to do, but I might be missing something

File: types/expressions.d.ts [hunk 4]
```diff
@@ -0,0 +1,2880 @@
...
+       * @see https://docs.mongodb.com/manual/reference/operator/aggregation/cos/#mongodb-expression-exp.-cos
+       */
+      $cos: NumberExpression;
+    }
+
+    export interface Tan {
+      /**
+       * Returns the tangent of a value that is measured in radians.
+       *
+       * @version 4.2
```
Comment (vkarpov15) [inline types/expressions.d.ts:186]: Can we get rid of the `@version` tags? It's a bit confusing because `@version` refers to the MongoDB server rather than the Mongoose version

**Comment by andreialecu** [comment]
Just a heads up that this PR broke a lot of our aggregations' typings, see more details in #11952
